lib/devicetrust/native/device_darwin.m

```diff
@@ -0,0 +1,249 @@
+#include "device_darwin.h"
+
+#import <CoreFoundation/CoreFoundation.h>
+#import <Foundation/Foundation.h>
+#import <IOKit/IOKitLib.h>
+#import <Security/Security.h>
+
+#include <stdlib.h>
+#include <string.h>
+
+const char *kDeviceKeyLabel = "com.gravitational.teleport.devicekey";
+
+const int kErrNoApplicationTag = 1;
+const int kErrNoValueRef = 2;
+const int kErrCopyPubKeyFailed = 3;
+const int kErrIOMatchingServiceFailed = 4;
+const int kErrIORegistryEntryFailed = 5;
+
+OSStatus findDeviceKey(NSNumber *retAttrs, CFTypeRef *out) {
+  NSData *label = [NSData dataWithBytes:(void *)kDeviceKeyLabel
+                                 length:strlen(kDeviceKeyLabel)];
+  NSDictionary *query = @{
+    (id)kSecClass : (id)kSecClassKey,
+    (id)kSecAttrKeyType : (id)kSecAttrKeyTypeECSECPrimeRandom,
+    (id)kSecMatchLimit : (id)kSecMatchLimitOne,
+    (id)kSecReturnRef : @YES,
+    (id)kSecReturnAttributes : (id)retAttrs,
+    (id)kSecAttrApplicationLabel : label,
+  };
+  return SecItemCopyMatching((CFDictionaryRef)query, out);
+}
+
+void copyPublicKey(const char *id, CFDataRef pubKeyRep, PublicKey *pubKeyOut) {
+  pubKeyOut->id = strdup(id);
+  pubKeyOut->pub_key_len = CFDataGetLength(pubKeyRep);
+  pubKeyOut->pub_key = calloc(pubKeyOut->pub_key_len, sizeof(uint8_t));
+  memcpy(pubKeyOut->pub_key, CFDataGetBytePtr(pubKeyRep),
+         pubKeyOut->pub_key_len);
+}
+
+int32_t DeviceKeyCreate(const char *newID, PublicKey *pubKeyOut) {
```

Comment:
It is a pity we don't have method comments here, only in headers. I wonder if there is a solution that could make it happen without duplicating the comments.

Reply:
I think that's par for the course for C - I would expect function docs to the at the .h, and implementation comment sprinkled through here.